# Введение в глубинное обучение, ФКН ВШЭ

## Домашнее задание 2. Классификация изображений. Сверточные нейронные сети.

### Общая информация

Дата выдачи: 07.11.2021

Мягкий дедлайн: 23:59MSK 05.12.2021

Жесткий дедлайн: 23:59MSK 05.12.2021

Оценка после штрафа после мягкого дедлайна вычисляется по формуле $M_{penalty} = M_{full} \cdot 0.85^{t/1440}$, где $M_{full}$ — полная оценка за работу без учета штрафа, а $t$ — время в минутах, прошедшее после мягкого дедлайна (округление до двух цифр после запятой). Таким образом, спустя первые сутки после мягкого дедлайна вы не можете получить оценку выше 8.5, а если сдать перед самым жестким дедлайном, то ваш максимум — 5.22 балла.

### Оценивание и штрафы

Максимально допустимая оценка за работу — 10 баллов. Сдавать задание после указанного срока сдачи нельзя.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Неэффективная реализация кода может негативно отразиться на оценке. Также оценка может быть снижена за плохо читаемый код и плохо оформленные графики. Все ответы должны сопровождаться кодом или комментариями о том, как они были получены.

### О задании

В этом задании потребуется обучить классификатор изображений. Будем работать с датасетом, название которого раскрывать не будем. Можете посмотреть самостоятельно на картинки, которые в есть датасете. В нём 200 классов и около 5 тысяч картинок на каждый класс. Классы пронумерованы, как нетрудно догадаться, от 0 до 199. Скачать датасет можно вот [тут](https://www.dropbox.com/s/33l8lp62rmvtx40/dataset.zip?dl=0).

Структура датасета простая -- есть директории train/ и val/, в которых лежат обучающие и валидационные данные. В train/ и val/ лежат директориии, соответствующие классам изображений, в которых лежат, собственно, сами изображения.

## Задание 1. (Максимум 10 баллов + 5 бонусных баллов)

__Необходимо выполнить любое из двух заданий (на выбор)__

1) Добейтесь accuracy **на валидации не менее 0.44**. В этом задании **запрещено** пользоваться предобученными моделями и ресайзом картинок. 

2) Добейтесь accuracy **на валидации не менее 0.84**. В этом задании делать ресайз и использовать претрейн можно. 

Обязательно указывайте ссылки на чужой код, если вы его используете. Обязательно ссылайтесь на статьи / блогпосты / вопросы на stackoverflow / видосы от ютуберов-машинлернеров / курсы / подсказки от Дяди Васи и прочие дополнительные материалы, если вы их используете. 

Ваш код обязательно должен проходить все `assert`'ы ниже.

Необходимо написать функции `train_one_epoch`, `train` и `predict` по шаблонам ниже (во многом повторяют примеры с семинаров). Обратите особое внимание на функцию `predict`: она должна возвращать список лоссов по всем объектам даталоадера, список предсказанных классов для каждого объекта из даталоалера и список настоящих классов для каждого объекта в даталоадере (и именно в таком порядке).

__Использовать внешние данные для обучения строго запрещено в обоих заданиях. Также запрещено обучаться на валидационной выборке__.


__Критерии оценки__: Оценка вычисляется по простой формуле: `min(10, 10 * Ваша accuracy / 0.44)` для первого задания и `min(10, 10 * (Ваша accuracy - 0.5) / 0.34)` для второго. Оценка округляется до десятых по арифметическим правилам. Если вы выполнили оба задания, то берется максимум из двух оценок.

__Бонус__. Вы получаете 5 бонусных баллов если справляетесь с обоими заданиями на 10 баллов (итого 15 баллов). В противном случае выставляется максимальная из двух оценок и ваш бонус равен нулю.

__Советы и указания__:
 - Наверняка вам потребуется много гуглить о классификации и о том, как заставить её работать. Это нормально, все гуглят. Но не забывайте, что нужно быть готовым за скатанный код отвечать :)
 - Используйте аугментации. Для этого пользуйтесь модулем `torchvision.transforms` или библиотекой [albumentations](https://github.com/albumentations-team/albumentations)
 - Можно обучать с нуля или файнтюнить (в зависимости от задания) модели из `torchvision`.
 - Рекомендуем написать вам сначала класс-датасет (или воспользоваться классом `ImageFolder`), который возвращает картинки и соответствующие им классы, а затем функции для трейна по шаблонам ниже. Однако делать это мы не заставляем. Если вам так неудобно, то можете писать код в удобном стиле. Однако учтите, что чрезмерное изменение нижеперечисленных шаблонов увеличит количество вопросов к вашему коду и повысит вероятность вызова на защиту :)
 - Валидируйте. Трекайте ошибки как можно раньше, чтобы не тратить время впустую.
 - Чтобы быстро отладить код, пробуйте обучаться на маленькой части датасета (скажем, 5-10 картинок просто чтобы убедиться что код запускается). Когда вы поняли, что смогли всё отдебажить, переходите обучению по всему датасету
 - На каждый запуск делайте ровно одно изменение в модели/аугментации/оптимайзере, чтобы понять, что и как влияет на результат.
 - Фиксируйте random seed.
 - Начинайте с простых моделей и постепенно переходите к сложным. Обучение лёгких моделей экономит много времени.
 - Ставьте расписание на learning rate. Уменьшайте его, когда лосс на валидации перестаёт убывать.
 - Советуем использовать GPU. Если у вас его нет, используйте google colab. Если вам неудобно его использовать на постоянной основе, напишите и отладьте весь код локально на CPU, а затем запустите уже написанный ноутбук в колабе. Авторское решение задания достигает требуемой точности в колабе за 45 минут обучения.
 
Good luck & have fun! :)

In [ ]:
import numpy as np
import torch
import torchvision
from tqdm import tqdm
import matplotlib.pyplot as plt
from torch import nn
from torch.nn import functional as F
from sklearn.metrics import accuracy_score
from torchvision.datasets import ImageFolder
import sys

In [ ]:
!wget https://www.dropbox.com/s/33l8lp62rmvtx40/dataset.zip?dl=1 -O dataset.zip && unzip -q dataset.zip

### Подготовка данных

In [ ]:
import random
def set_random_seed(seed):
    torch.backends.cudnn.deterministic = True
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)

In [ ]:
from torchvision.transforms import Normalize, ToTensor, Compose, ColorJitter

In [ ]:
set_random_seed(101)

In [ ]:
transform=Compose([ToTensor(), Normalize((0.5, 0.5, 0.5), (1, 1, 1))])
train_transform = transform 
val_transform = transform

train_dataset = ImageFolder("./dataset/dataset/train", transform=train_transform)
val_dataset = ImageFolder("./dataset/dataset/val", transform=val_transform)

train_dataloader = torch.utils.data.DataLoader(train_dataset, batch_size=256, shuffle=True)
val_dataloader = torch.utils.data.DataLoader(val_dataset, batch_size=256, shuffle=False)

In [ ]:
# Just very simple sanity checks
assert isinstance(train_dataset[0], tuple)
assert len(train_dataset[0]) == 2
assert isinstance(train_dataset[1][1], int)
print("tests passed")

tests passed


### Вспомогательные функции, реализация модели

In [ ]:
#тут использовался код с семинаров, evaluate ниже тоже с семинаров
def train_one_epoch(model, data_loader, criterion, optimizer, device="cuda:0", return_losses=False):
    model = model.to(device).train()
    total_loss = 0
    num_batches = 0
    all_losses = []
    total_predictions = np.array([])
    total_labels = np.array([])
    with tqdm(total=len(data_loader), file=sys.stdout) as prbar:
        for images, labels in data_loader:
            # Move Batch to GPU
            images = images.to(device)
            labels = labels.to(device)
            predicted = model(images)
            loss = criterion(predicted, labels)
            # Update weights
            loss.backward()
            optimizer.step()
            optimizer.zero_grad()
            # Update descirption for tqdm
            accuracy = (predicted.argmax(1) == labels).float().mean()
            prbar.set_description(
                f"Loss: {round(loss.item(), 4)} "
                f"Accuracy: {round(accuracy.item(), 4)}"
            )
            prbar.update(1)
            total_loss += loss.item()
            total_predictions = np.append(total_predictions, predicted.argmax(1).cpu().detach().numpy())
            total_labels = np.append(total_labels, labels.cpu().detach().numpy())
            num_batches += 1
            all_losses.append(loss.detach().item())
    metrics = {"loss": total_loss / num_batches}
    metrics.update({"accuracy": accuracy_score(total_predictions, total_labels)})
    if return_losses:
        return metrics, all_losses
    else:
        return metrics


def predict(model, data_loader, criterion, device="cuda:0"):
    model = model.eval()
    total_loss = 0
    num_batches = 0
    predicted_classes = np.array([])
    true_classes = np.array([])
    losses = np.array([])
    with tqdm(total=len(data_loader), file=sys.stdout) as prbar:
        for images, labels in data_loader:
            images = images.to(device)
            labels = labels.to(device)
            predicted = model(images)
            loss = criterion(predicted, labels)
            accuracy = (predicted.argmax(1) == labels).float().mean()
            prbar.set_description(
                f"Loss: {round(loss.item(), 4)} "
                f"Accuracy: {round(accuracy.item(), 4)}"
            )
            prbar.update(1)
            total_loss += loss.item()
            losses = np.append(losses, loss.item())  
            predicted_classes = np.append(predicted_classes, predicted.argmax(1).cpu().detach().numpy())
            true_classes = np.append(true_classes, labels.cpu().detach().numpy())
            num_batches += 1
    metrics = {"loss": total_loss / num_batches}
    metrics.update({"accuracy": accuracy_score(predicted_classes, true_classes)})

    return losses, predicted_classes, true_classes


def train(model, train_dataloader, val_dataloader, criterion, optimizer, device="cuda:0", epochs=10, scheduler=None):
    all_train_losses = []
    epoch_train_losses = []
    epoch_eval_losses = []
    for epoch in range(epochs):
        # Train step
        print(f"Train Epoch: {epoch}")
        train_metrics, one_epoch_train_losses = train_one_epoch(
            model=model,
            data_loader=train_dataloader,
            optimizer=optimizer,
            return_losses=True,
            criterion=criterion,
        )
        # Save Train losses
        all_train_losses.extend(one_epoch_train_losses)
        epoch_train_losses.append(train_metrics["loss"])
        # Eval step
        print(f"Validation Epoch: {epoch}")
        with torch.no_grad():
            all_losses, predicted_labels, true_labels = predict(model, val_dataloader, criterion, device)

### Обучение модели, запуски экспериментов

In [ ]:
torch.cuda.is_available()

True

In [ ]:
from torchvision.models import resnet34, resnet18

In [ ]:
set_random_seed(101)

model = resnet34(pretrained=False)
optimizer = torch.optim.Adam(model.parameters(), 1e-3)
criterion = nn.CrossEntropyLoss()
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=2)
n_epochs = 7
device = torch.device("cuda:0")

In [ ]:
model = model.cuda()

Простой тест на проверку правильности написанного кода

In [ ]:
all_losses, predicted_labels, true_labels = predict(model, val_dataloader, criterion, device)
assert len(predicted_labels) == len(val_dataset)
accuracy = accuracy_score(predicted_labels, true_labels)
print("tests passed")

Loss: 6.7709 Accuracy: 0.0: 100%|██████████| 40/40 [00:11<00:00,  3.42it/s]
tests passed


Запустить обучение можно в ячейке ниже.

In [ ]:
train(model, train_dataloader, val_dataloader, criterion, optimizer, device, n_epochs, scheduler)

Train Epoch: 0
Loss: 4.0957 Accuracy: 0.1125: 100%|██████████| 391/391 [04:14<00:00,  1.54it/s]
Validation Epoch: 0
Loss: 3.3453 Accuracy: 0.3125: 100%|██████████| 40/40 [00:10<00:00,  3.64it/s]
Train Epoch: 1
Loss: 3.397 Accuracy: 0.2125: 100%|██████████| 391/391 [04:12<00:00,  1.55it/s]
Validation Epoch: 1
Loss: 3.9631 Accuracy: 0.0625: 100%|██████████| 40/40 [00:11<00:00,  3.61it/s]
Train Epoch: 2
Loss: 2.9395 Accuracy: 0.3: 100%|██████████| 391/391 [04:13<00:00,  1.54it/s]
Validation Epoch: 2
Loss: 2.7836 Accuracy: 0.1875: 100%|██████████| 40/40 [00:11<00:00,  3.60it/s]
Train Epoch: 3
Loss: 2.9651 Accuracy: 0.3063: 100%|██████████| 391/391 [04:12<00:00,  1.55it/s]
Validation Epoch: 3
Loss: 2.6434 Accuracy: 0.3125: 100%|██████████| 40/40 [00:11<00:00,  3.57it/s]
Train Epoch: 4
Loss: 2.675 Accuracy: 0.35: 100%|██████████| 391/391 [04:14<00:00,  1.54it/s]
Validation Epoch: 4
Loss: 2.116 Accuracy: 0.4375: 100%|██████████| 40/40 [00:11<00:00,  3.55it/s]
Train Epoch: 5
Loss: 2.3875 Accur

### Проверка полученной accuracy

После всех экспериментов которые вы проделали, выберите лучшую из своих моделей, реализуйте и запустите функцию `evaluate`. Эта функция должна брать на вход модель и даталоадер с валидационными данными и возврашать accuracy, посчитанную на этом датасете.

In [ ]:
def evaluate(model, eval_dataloader, criterion, device="cuda:0"):
    cumulative_loss = 0
    acc = 0
    model = model.to(device).eval()
    with torch.no_grad():
        for idx, (images, labels) in enumerate(eval_dataloader): 
            images, labels = images.to(device), labels.to(device)
            preds = model(images)
            acc += (preds.argmax(1) == labels).float().mean()
    print("accuracy = {:.4f}".format(acc / idx))
    return acc / idx

In [ ]:
acc = evaluate(model, val_dataloader, criterion)
print(acc)

accuracy = 0.3633
tensor(0.3633, device='cuda:0')


In [ ]:
all_losses, predicted_labels, true_labels = predict(model, val_dataloader, criterion, device)
assert len(predicted_labels) == len(val_dataset)
accuracy = accuracy_score(true_labels, predicted_labels)
print(accuracy)
print(f'Оценка за 1 задание составит {np.clip(10 * accuracy / 0.44, 0, 10):.2f} баллов.')

Loss: 1.7183 Accuracy: 0.6875: 100%|██████████| 40/40 [00:11<00:00,  3.60it/s]
0.3462
Оценка за 1 задание составит 7.87 баллов.


## Задание 2 (0 баллов, но при невыполнении максимум за все задание — 0 баллов)

Напишите небольшой отчет о том, как вы добились полученного качества: какие средства использовали и какие эксперименты проводили. Подробно расскажите об архитектурах и значениях гиперпараметров, а также какие метрики на тесте они показывали. Чтобы отчет был зачтен, необходимо привести хотя бы 3 эксперимента.

Я решала первую задачу, используя готовые не предобученные сети.

## Эксперимент 1 (итоговый):
Использовался только Normalize((0.5, 0.5, 0.5), (1, 1, 1))

batch size : 256

Архитектура: resnet34, без предтрейна

Оптимизатор: Adam, lr = 1e-3

Функция потерь: CrossEntropyLoss

scheduler: ReduceLROnPlateau

Кол-во эпох: 7

accuracy = 0.3462

Если увеличить число эпох сеть сильно переобучалась в данном случае.

## Эксперимент 2:

Использовался только Normalize((0.5, 0.5, 0.5), (1, 1, 1))

batch size : 256

Архитектура: resnet18, без предтрейна

Оптимизатор: Adam, lr = 3e-3

Функция потерь: CrossEntropyLoss

scheduler: ReduceLROnPlateau

Кол-во эпох: 7

accuracy : 0.3278

## Эксперимент 3:
Normalize((0.5, 0.5, 0.5), (1, 1, 1)), ColorJitter(brightness=.3, hue=.2)

batch_size = 256

Архитектура: resnet18, без предтрейна

Оптимизатор: Adam, lr = 3e-3

Функция потерь: CrossEntropyLoss

scheduler: не используется

Кол-во эпох: 11

accuracy: 0.2437

Самый неудачный эксперимент. Причем увеличить число эпох особо не помогло, модель только начала переобучаться.